### DISCUSSION TITLE
How can I read 4B edge graph into Networkit fast?

### DISCUSSION THREAD

**Discussion by Ishitori** [networkit/networkit#1040] [Q&A] [answered]
Hi there,

I have a 230M nodes and 4B edge graph (non-directed, so each edge is duplicated twice, weighted), and I want to apply Leiden community detection algorithm. I store data in a set of .parquet files: src_id, dst_id, weight.

1. What is the easiest way to read it by Networkit? if it helps, I have code which can read it all and convert to METIS list of arrays xadj, adjncy, adjwgt (in ctypes). Can I use these directly to construct a graph in Python, without relying on readers?
2. Can Networkit actually handle a graph of that size?

**Comment by hmeyerhenke** [comment, marked as answer]
Hello Ishitori,

This size should really be no problem if you have enough memory (I estimate roughly 100 GB). We have applied community detection algorithms to graphs of that size/order before (and larger ones) on compute servers with sufficient RAM.

NetworKit supports a variety of graph formats. Please consult the documentation if something comes close to your format.

What I can say right now: What we call the METIS format is well supported. However, for graphs of this size, reading the data takes quite some time. Which is true for all text formats, this is not a problem of the METIS format per se and is likely to be true for your format as well.

My suggestion would thus be to convert your large graphs into NetworKit's binary format. Reading in those graphs is much faster. Of course, reading the graphs in once from the old format to perform the conversion is a necessity, but at least I/O will be much faster afterwards.

Hope this helps,
Henning

**Reply by Ishitori** [reply]
Thanks for the reply, Henning. 

I have enough RAM to fit the graph in memory, thanks for pointing that out. 

Do you know from the top of your head if any formats support a graph storage in multiple files? My graph is an output of Spark job, so it is a few hundreds files, which I would like to pass as-is to Networkit. I can `cat` it into a single file, but that would take twice as much of disk space. Any suggestions?

**Reply by hmeyerhenke** [reply]
To be honest, I am not 100% sure, but I doubt that multiple files can be handled this way in the current codebase. But there are people who know the code better than me... You are probably best off with cat and a conversion to binary. In the long run that saves space and time.

**Reply by fabratu** [reply]
I can confirm, that we currently do not support reading a graph from multiple files. This is on the agenda of nice-to-have features though. If there is enough memory, you could parse the output files from Spark to a RAM disk. The binary format not only saves time for IO, it also consumes less space after conversion.

**Reply by adsharma** [reply]
Please take a look at https://github.com/Ladybug-Memory/icebug-format

This can be loaded efficiently into Apache Arrow backed memory and processed more efficiently than a `std::vector` based mutable Graph implementation.

I'm maintaining a fork here: https://github.com/Ladybug-Memory/icebug
